### ISSUE TITLE
DaemonV2

### ISSUE THREAD

**Issue by terrycain** [lah7/razer-chroma-drivers#19]
You mind testing my new daemon soonish?

Would also be useful as you'll need to adapt the program to work with the new DBus interface (though I plan to provide a nice python interface)

Also good news, I have fixed the macro_keys FN+F1-F7 not working (well fixed in software at least)

**Comment by lah7** [comment]
Sure. I gave your `DaemonV2` branch a test by compiling and installing as a package (and `sudo make install`) ... it didn't go well.

* The service still worked, so tray and GUI still works. Nothing broke there.
* `daemon2/run_razer_daemon.py` just exits every time. I'm not sure what's happening... it only outputs this:

```
/usr/lib/python3/dist-packages/razer/keyboard.py:5: PyGIWarning: Gdk was imported without specifying a version first. Use gi.require_version('Gdk', '3.0') before import to ensure that the right version gets loaded.
  from gi.repository import Gdk
```

Harmless message of course. I opened up to discover a foreground flag... that's busted:
```
Traceback (most recent call last):
  File "/tmp/razer_drivers-daemonV2/daemon2/run_razer_daemon.py", line 48, in <module>
    run()
  File "/tmp/razer_drivers-daemonV2/daemon2/run_razer_daemon.py", line 42, in run
    daemonize(**daemon_args)
  File "/tmp/razer_drivers-daemonV2/daemon2/razer_daemon/daemon.py", line 101, in daemonize
    daemon = RazerDaemon(verbose, log_dir, console_log, run_dir, config_file)
  File "/tmp/razer_drivers-daemonV2/daemon2/razer_daemon/daemon.py", line 155, in __init__
    log_file = os.path.join(log_dir, 'razer.log')
  File "/usr/lib/python3.5/posixpath.py", line 89, in join
    genericpath._check_arg_types('join', a, *p)
  File "/usr/lib/python3.5/genericpath.py", line 143, in _check_arg_types
    (funcname, s.__class__.__name__)) from None
TypeError: join() argument must be str or bytes, not 'NoneType'
```

* I tried `daemon2/razer_daemon/misc/ripple_effect.py`, but my cursor changes to a cross hair... click a few times and then it quits.
```
/tmp/razer_drivers-daemonV2/daemon2/razer_daemon/misc/ripple_effect.py: line 3: $'\nContains the functions and classes to perform ripple effects\n': command not found
from: too many arguments
/tmp/razer_drivers-daemonV2/daemon2/razer_daemon/misc/ripple_effect.py: line 13: syntax error near unexpected token `('
/tmp/razer_drivers-daemonV2/daemon2/razer_daemon/misc/ripple_effect.py: line 13: `class RippleEffectThread(threading.Thread):'
```

* Same happens with `macro.py`. It leaves some screenshots behind.... I think I've seen something like that happen before.

-------
So... I don't think it's installed or running properly. I can't confirm that the software fix for the activated macro keys and FN F1-F7 either. :confused: I'm probably doing something wrong?

---------
If you go ahead with a new Python interface, please do document it as you go along (on the wiki maybe?). :smile:

**Comment by terrycain** [comment]
its not ready yet. It will be soon :wink: currently the makefile will only install the old daemon which I haven't touched.

Give me a few days and ill get it to a workable state and make some instructions on how to get it running. I've decided instead of making a system daemon this one will be ran in userspace so that it can access the users DBus to detect screensaver stuff. 

I'll document the python interface once I've got it working. It'll expose each device on DBus by its serial, you can use d-feet (apt install it) to investigate DBus.

**Comment by lah7** [comment]
I just re-read this bit:

> You mind testing my new daemon soonish?

"Soonish"... d'oh! 😫 I originally thought "soon" today! Oh dear...

---

Good idea for going user space... I've been wondering about a "matrix" effect when my screensaver starts. :wink:

**Comment by terrycain** [comment]
> Good idea for going user space...

Cheers, I though it was beneficial just need to look into how I want to manage it, I know upstart and systemd allow user jobs to be stored in ~/.config but I'm thinking just to set it up in /etc/init or /etc/systemd and not set it to autostart and then set it to autostart on logon (like how the tray applet is supposed to).

> "Soonish"... d'oh!

No worries. When it's running it works quite well but needs some stuff set up like my udev script setting the files in the driver directory to the `plugdev` group (as well as `/dev/input/event*` to `plugdev`) and the user being added to `plugdev`

> I've been wondering about a "matrix" ... I imagine a key framed "lighting effects" editor

I have thought about this recently and definitely want something like it. If you build it into the app then I'll support it. Another thing if the effect you make repeats itself we could compile it to binary the driver accepts so it wouldn't have to be computed on the fly :). Definitely need to thing how I'm going to do it though.

I make a custom wrapper for DBus services (acting as the server) and I want to improve that once I've got the daemon stable as currently it doesn't support method overloading which would come in handy.

I will get some docs soon enough but if it goes to plan, shouldn't be too difficult.

Once the new driver is in place it _should_ make things more stable and _should_ do away with the ugly startup script that binds things.

Another thing I'll need you to test soon is my udev script. It works for me but it'll need thorough testing.

**Comment by lah7** [comment]
It all sounds good, but on second thoughts, a timeline "key frame" type of editor for effects sounds like a future challenge.

After the new daemon and GUI is up to scratch, I'd quite like to be able to set effects per key for profiles, and then the fancy effect editor can come later. :wink: Like as discussed here: https://github.com/lah7/razer-chroma-drivers/issues/8

I suppose this is a lot more accomplishable with the new daemon :smile:

**Comment by terrycain** [comment]
Yep, I think I'm done on the features for now. Just need to test it some more and package it up. Then the next thing on the list is make the first stable merge and then draw up a nice plan.

**Comment by terrycain** [comment]
@lah7 Can you test my udev script in my daemonV2 branch. Ideally remove the current udev script your using, add in my test one. Stop the daemon from starting at boot (ideally remove it just to be sure but stopping it starting should work). Then reboot and ideally the devices should be bound at boot. And then test unplugging and repluggin them and see if they are still bound.

Could you also check permissions in the device directory too, files should be owned by root:plugdev

**Comment by lah7** [comment]
Here's what I did:

* Reinstalled the driver (via `package_for_debian.sh`)

The service didn't run after restarting the computer:
```
May 26 12:48:28 LukeUbuntuDesktop systemd[1]: Starting Razer Keyboard Daemon...
May 26 12:48:28 LukeUbuntuDesktop systemd_helpers.sh[895]: Unbinding device (0003:1532:0203.0004) from hid-generic
May 26 12:48:28 LukeUbuntuDesktop systemd[1]: razer_bcd.service: Control process exited, code=exited status=2
May 26 12:48:28 LukeUbuntuDesktop systemd_helpers.sh[895]: Binding device (0003:1532:0203.0004) to razerkbd
May 26 12:48:28 LukeUbuntuDesktop systemd_helpers.sh[895]: /usr/share/razer_bcd/systemd_helpers.sh: 22: /usr/share/razer_bcd/systemd_helpers.sh: cannot create /sys/bus/hid/drivers/razerkbd/bind: Directory nonexi
May 26 12:48:28 LukeUbuntuDesktop systemd[1]: Failed to start Razer Keyboard Daemon.
May 26 12:48:29 LukeUbuntuDesktop systemd[1]: razer_bcd.service: Unit entered failed state.
May 26 12:48:29 LukeUbuntuDesktop systemd[1]: razer_bcd.service: Failed with result 'exit-code'.
```

But it worked after manually issuing `sudo service razer_bcd restart`:
```
May 26 12:55:04 LukeUbuntuDesktop systemd[1]: Starting Razer Keyboard Daemon...
May 26 12:55:04 LukeUbuntuDesktop systemd_helpers.sh[3175]: Device already binded
May 26 12:55:04 LukeUbuntuDesktop razer_bcd[3215]: Starting razer blackwidow chroma daemon as a daemon
May 26 12:55:04 LukeUbuntuDesktop systemd[1]: Started Razer Keyboard Daemon.
```

`99-razer.rules` is present in `/etc/udev/rules.d/99-razer.rules`.

Running `daemon2/run_razer_daemon.py` ran then quit (just a PyGIWarning)... but the ripple effect script did not work. Re-plugging in the device didn't change anything nor could be controlled by the GUI.

For the device blocks, all of them are still owned by `root`, `plugdev` is a group with read/write access.

----------
I'm not sure if I'm misunderstanding how this is working?

**Comment by terrycain** [comment]
I just wanted the udev part tested (only needed copying the udev rules).

Ok, I've generated some debs (the ./package_for_launchpad.sh script made these but I havent made any docs for making these debs yet), they are [here](https://drive.google.com/folderview?id=0BydYgD2LzDUCeHR4SnJFN25BeFU&usp=sharing). 

If you wouldn't mind, can you remove the driver and daemon and install the debs in the link (driver first, then python3-razer then razer-daemon). Then reboot. When you log in it should have started. Get the app d-feet, open it, click on the session bus tab, and look for `org.razer` in the list.

If its not there try `status razer-service`. Also there might be log files in `~/.razer-service/logs`

**Comment by lah7** [comment]
I'm happy to give things a test run -- providing I know what I'm doing of course. :wink:

I'll be switching back to the old one after testing so I can continue working on [polychromatic](https://github.com/lah7/polychromatic). Packages make that easy. :relieved: 

----------
So, after uninstalling the old daemon installing the new debs, then giving the system a reboot... I can't seem to find `org.razer` in d-feet... nor can I find a service or init script, but there is a `~/.razer-service` folder with two empty folders inside: `data` and `logs` , so there's a start. Not seen `/tmp/razer_udev.log` before either.

---------
`status razer-service` won't work since I'm running Ubuntu MATE 16.04, which is systemd, not upstart. There were no systemd service either.

----------
However, I did get it work by running:

    python3 /usr/lib/python3/dist-packages/razer_daemon/daemon.py

But first I needed to create the log directories or else it failed:

    mkdir -p /tmp/razer_daemon/logs

Then `org.razer` appeared in the session bus.

This error was looping constantly:
```
2016-05-26 21:35:01 | razer.screensaver              | INFO     | Initialising DBus screensaver object
2016-05-26 21:35:01 | razer.screensaver              | ERROR    | Caught exception whilst trying to get screensaver DBus
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/dbus/bus.py", line 175, in activate_name_owner
    return self.get_name_owner(bus_name)
  File "/usr/lib/python3/dist-packages/dbus/bus.py", line 361, in get_name_owner
    's', (bus_name,), **keywords)
  File "/usr/lib/python3/dist-packages/dbus/connection.py", line 651, in call_blocking
    message, timeout)
dbus.exceptions.DBusException: org.freedesktop.DBus.Error.NameHasNoOwner: Could not get owner of name 'com.canonical.Unity': no such name

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/razer_daemon/misc/screensaver_thread.py", line 37, in load_dbus
    unity_object = session_bus.get_object('com.canonical.Unity', '/org/gnome/ScreenSaver')
  File "/usr/lib/python3/dist-packages/dbus/bus.py", line 241, in get_object
    follow_name_owner_changes=follow_name_owner_changes)
  File "/usr/lib/python3/dist-packages/dbus/proxies.py", line 248, in __init__
    self._named_service = conn.activate_name_owner(bus_name)
  File "/usr/lib/python3/dist-packages/dbus/bus.py", line 180, in activate_name_owner
    self.start_service_by_name(bus_name)
  File "/usr/lib/python3/dist-packages/dbus/bus.py", line 278, in start_service_by_name
    'su', (bus_name, flags)))
  File "/usr/lib/python3/dist-packages/dbus/connection.py", line 651, in call_blocking
    message, timeout)
dbus.exceptions.DBusException: org.freedesktop.DBus.Error.ServiceUnknown: The name com.canonical.Unity was not provided by any .service files
```

Most likely because the MATE desktop does not have these Unity things.

Other scripts in `/usr/lib/python3/dist-packages/razer_daemon/misc` didn't do anything either. 

------------
> When you log in it should have started.

In MATE (I think LXDE too), a `.desktop` file can be placed in `~/.config/autostart` for it to auto start on logs on. I believe you're running Ubuntu 14.04 "Unity"?

**Comment by terrycain** [comment]
Run `ls -l /usr/share/upstart/sessions/` as I know on my laptop 15.04 (which is systemd) Upstart lives on as a session manager (Upstart ftw). Am pretty sure Ubuntu uses Upstart to run user sessions and indicatior applets etc... Upstart has this really nice concept of session jobs which are run once per session (and most importantly the DBus env variable will be set without any hacks, as this variable is key and also the devil).

I had the issue with it not finding the screensaver too because when it starts it checks before the screensaver exists (which is a similar but separate problem), I have got a fix for that I'm working on.

For starters running `daemon.py` is not recommended running `/usr/sbin/razer-service` is better (you can add `--foreground` to not daemonise). `man razer-service` and `man razer.conf` should help. Files inside misc are not scripts, they are misc classes used by the dameon (I just cba to find a proper place for them yet).

Basically it crashes on the screensaver part. My new version (not ready yet) fixes that.

On Unity there is
```
Bus name: com.canonical.Unity
Object path: /org/gnome/ScreenSaver
Interface: org.gnome.ScreenSaver
Method: GetActive() -> Boolean
```

What is the equivalent on your pc? Will take a little bit of DBus investigation (make sure its on the session bus)

**Comment by terrycain** [comment]
Also `/tmp/razer_udev.log` is a log file written to when my udev rules script called a mount script to do binding etc... See `/lib/udev/razer_mount` I think its called for more details. I might remove it but currently leaving it in is doing no harm.

And that's the benefit of packaging, easy install, easy uninstall, easy upgrade :smile:.

> I'm happy to give things a test run -- providing I know what I'm doing of course. 
Yeah, there's not much to it bar getting it running at the moment, I need to make a library for your app to interface with DBus nicely. If you have any idea what sort of API you want then I'm all ears, otherwise I'll just try and make it as useful as possible.

**Comment by lah7** [comment]
> Run ls -l /usr/share/upstart/sessions/

Hmm, and so it is! There's a bunch of `conf` files in there.

----------
For the screensaver, MATE is very similar to GNOME, it's under here:
`org.mate.ScreenSaver`
`GetActive() -> Boolean`

-------
I'll give the new daemon another test when your fixes are ready.

Next step is to add a new class for multi-device support in Polychromatic, then re-juggle the UI so it's dynamic depending what the device needs. As I'm juggling a few projects, I'm :snail: 's pace...

**Comment by terrycain** [comment]
#Upstart :smile: (much prefer upstart personally), The new daemon does install itself there thought :wink: so it will get started after DBus.

----

I need the bus path, object path and interface name to code in the DBus screensaver stuff. I assume its just 
```
Bus Name: org.mate.ScreenSaver
Object path: /org/gnome/ScreenSaver
Interface name: org.gnome.ScreenSaver
GetActive() -> Boolean
```
?

---

Wicked, yeah I'm away from home so can't test anything for a week or two but can at least get it working for you. Am currently writing unit tests :neutral_face: as thought I might as well.

Once I've got the driver working for you, I'll work on the library to interface with it. 

And I've got no problem with :snail:'s pace. What else you working on? anything interesting?

**Comment by lah7** [comment]
Close:
```
Bus Name: org.mate.ScreenSaver
Object path: /org/mate/ScreenSaver
Interface name: org.mate.ScreenSaver
GetActive() -> Boolean
```

-------
I work on a couple of projects for Ubuntu MATE, but they're on BitBucket, not GitHub. I do enjoy working on them, but of course I can only work on one thing at a time! :smile:

In the next day or so, I'll push some commits for Polychromatic's multi-device implementation. It will give you the idea how this "middle ground" will work between the UI and daemon calls when you come to designing the new library. A big plus if it's capable of detecting/remembering the "current effect" and colours in use (if applicable). :wink:

**Comment by terrycain** [comment]
The keyboard isn't capable of remembering the current effect. There are some edge cases where it is different. If its a non-chroma it can tell what effect is being used. Like my BW2013 setting the effect on a logo (effects the entire keyboard) that effect can be detected (as technically its not classed as a matrix effect). 

I could add in some code to remember the last active effect but it will get out of sync if someone activates  effects via the driver directly

**Comment by lah7** [comment]
I see. This sounds like something better handled by the Polychormatic libraries, seeing as it is the GUI / tray that would benefit having the "last effect" remembered.

**Comment by terrycain** [comment]
Yeah, if you want to keep track of colours and effects that's fine by me. It would be more suited to your app anyway as you can have a warning saying if you mess with things its your fault if things get out of sync.

Also new debs available [here](https://drive.google.com/folderview?id=0BydYgD2LzDUCeHR4SnJFN25BeFU&usp=sharing). _Should_ work with your DBus screensaver.

Best way to test, if it doesn't start up that is, would be to run `/usr/sbin/razer-service --verbose --foreground`.

Also do you reckon anything's missing from `man 8 razer-service` and `man 5 razer.conf`

**Comment by lah7** [comment]
Installed, restarted, again, it did not start automatically - no `razer-service` process to be seen. Running it in the foreground works.

No stdout spam complaining about the screensaver :+1: 

------------
I have noticed there is supposed to be a `~/.razer-service/razer.conf` file, but it's missing.

--------
The man pages look fine to me :+1: - I did spot a typo in `razer-service(8)`: _immediatly_

-------
I re-plugged the keyboard to test, this happened:
```
Exception in thread Thread-2:
Traceback (most recent call last):
  File "/usr/lib/python3.5/threading.py", line 914, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3/dist-packages/razer_daemon/misc/key_event_management.py", line 167, in run
    key_data = event_file.read(EVENT_SIZE)
OSError: [Errno 19] No such device
```

But the daemon didn't crash, so all is good.

----
Tried `d-feet` to set lighting, most worked, but a few just "froze" on the keyboard. Ripple and Reactive "didn't work", possibly because it froze somewhere and didn't want to quit:

```
^C2016-05-29 11:35:34 | razer                          | INFO     | Stopping daemon.
2016-05-29 11:35:34 | razer.screensaver              | INFO     | Screensaver Thread finished
2016-05-29 11:35:34 | razer.device0.keymanager       | DEBUG    | Stopping key manager
2016-05-29 11:35:34 | razer.device0.ripplemanager    | DEBUG    | Closing Ripple Manager
^C2016-05-29 11:35:35 | razer                          | INFO     | Stopping daemon.
2016-05-29 11:35:36 | razer.device0.ripplemanager    | ERROR    | Could not stop RippleEffect thread
^C2016-05-29 11:35:43 | razer                          | INFO     | Stopping daemon.
^C2016-05-29 11:37:01 | razer                          | INFO     | Stopping daemon.
^C2016-05-29 11:37:01 | razer                          | INFO     | Stopping daemon.
```

Then after CTRL+Z'd out and re-plugging (bad idea?), the daemon and dbus was no more:
```
Traceback (most recent call last):
  File "/usr/sbin/razer-service", line 64, in <module>
    run()
  File "/usr/sbin/razer-service", line 58, in run
    daemonize(**daemon_args)
  File "/usr/lib/python3/dist-packages/razer_daemon/daemon.py", line 106, in daemonize
    daemon = RazerDaemon(verbose, log_dir, console_log, run_dir, config_file)
  File "/usr/lib/python3/dist-packages/razer_daemon/daemon.py", line 176, in __init__
    self._load_devices()
  File "/usr/lib/python3/dist-packages/razer_daemon/daemon.py", line 288, in _load_devices
    razer_device = device_class(device_path, device_number, self._config)
  File "/usr/lib/python3/dist-packages/razer_daemon/hardware/keyboards.py", line 61, in __init__
    super(RazerBlackWidowChroma, self).__init__(*args)
  File "/usr/lib/python3/dist-packages/razer_daemon/hardware/keyboards.py", line 19, in __init__
    super(MacroKeyboard, self).__init__(*args)
  File "/usr/lib/python3/dist-packages/razer_daemon/hardware/device_base.py", line 55, in __init__
    DBusService.__init__(self, self.BUS_PATH, object_path)
  File "/usr/lib/python3/dist-packages/razer_daemon/dbus_services/service.py", line 85, in __init__
    super(DBusService, self).__init__(bus_object, object_path)
  File "/usr/lib/python3/dist-packages/dbus/service.py", line 455, in __init__
    validate_object_path(object_path)
ValueError: Invalid object path '/org/razer/device/': ends with '/' and is not just '/'
```
It looked like it was working again a minute later, but DBUS wasn't happy in `d-feet`. :-1:

**Comment by terrycain** [comment]
Hmm, not sure why its not starting. Could you stop it, delete ~/.razer-service, then reboot? If it reappears then at least its trying to start. (eventually when I get home I will look at ubuntu with MATE to see whats going on).

---

It might not be copying `razer.conf` from `/usr/share/razer-service/razer.conf.example`.

---

Ta, fixed the typo.

---

Yeah, basically the daemon spam reads event files in `/dev/input`, whilst originally the daemon handled unplugging and repluging, when I started adding other bits, I need to handle that again. What happens is there is a thread that watches the keys (from the event file), when it got the exception, that thread died but as its a thread the rest of the daemon remains ok. The ripple effect thread probably was effected when the key watching thread died (if that died first, otherwise not sure). As the ripple effect requires to know what keys are doing what in order to draw circles.

Tis strange reactive didn't work. Hmm.

---

If you Ctrl+Z'd it, it would of entered paused state which isnt good for threaded programs when things it uses (like event files, as they would of reappeared but with different file handles) change. 

I get that error sometimes randomly when I restart the daemon when debugging, but I can't seem to find out what causes it as its intermittent.

---

Least its detecting the screensaver :).

Ill work on fixing the event file bit. The solution is a little hacky to begin with :smile:. It started out using epoll to watch the file and read it only when events were written to it which worked, but epoll was only telling me there was an event when another event was triggered (so the latest event would not be triggered until another event was fired). This caused major annoyance (only because of the bloody FN key not treated as a modifier key). So switched to a non-blocking read in a loop super fast with a sleep in there which is more inefficient but results in a better experience.

**Comment by lah7** [comment]
Sorry, I should've replied to you earlier!

-----
> Could you stop it, delete ~/.razer-service, then reboot?

Tried this, still nowhere to be seen. When starting it again manually this time, `razer.conf` was present.

The easiest solution for the MATE desktop is to add a `.desktop` file here:
`~/.config/autostart/razer-service.desktop`

Containing:
```
[Desktop Entry]
Type=Application
Exec=razer-service
Hidden=false
X-MATE-Autostart-enabled=true
Name=Razer Service
Comment=
```

That's the file created when added with the **Start-up Programs** program. I **think** LXDE uses the same place for auto-started user programs.

-----------
I got to see one ripple this time (`0,255,0,0.100`), but then it freezes a second after. Restarted and tried again (`0,2550,0.001`), and to my surprise, **it works**! Wicked! So the refresh rate parameter is accountable.

This did pop up and prevented CTRL+C... (I was still able to execute another service in the foreground without any trouble)
```
Traceback (most recent call last):
  File "/usr/sbin/razer-service", line 64, in <module>
    run()
  File "/usr/sbin/razer-service", line 58, in run
    daemonize(**daemon_args)
  File "/usr/lib/python3/dist-packages/razer_daemon/daemon.py", line 106, in daemonize
    daemon = RazerDaemon(verbose, log_dir, console_log, run_dir, config_file)
  File "/usr/lib/python3/dist-packages/razer_daemon/daemon.py", line 176, in __init__
    self._load_devices()
  File "/usr/lib/python3/dist-packages/razer_daemon/daemon.py", line 288, in _load_devices
    razer_device = device_class(device_path, device_number, self._config)
  File "/usr/lib/python3/dist-packages/razer_daemon/hardware/keyboards.py", line 61, in __init__
    super(RazerBlackWidowChroma, self).__init__(*args)
  File "/usr/lib/python3/dist-packages/razer_daemon/hardware/keyboards.py", line 19, in __init__
    super(MacroKeyboard, self).__init__(*args)
  File "/usr/lib/python3/dist-packages/razer_daemon/hardware/device_base.py", line 55, in __init__
    DBusService.__init__(self, self.BUS_PATH, object_path)
  File "/usr/lib/python3/dist-packages/razer_daemon/dbus_services/service.py", line 85, in __init__
    super(DBusService, self).__init__(bus_object, object_path)
  File "/usr/lib/python3/dist-packages/dbus/service.py", line 455, in __init__
    validate_object_path(object_path)
ValueError: Invalid object path '/org/razer/device/': ends with '/' and is not just '/'
```

I found changing between **ripple → wave** caused a "freeze", but setting **ripple → static → wave** seemed OK. Reactive is broken somewhere, nothing happens.

I let the screen saver come on... success! The keyboard turned off and back on when the mouse moved. :smile: I'm wondering now if it's possible to set a custom effect or run a script when that happens... might be a nice option for the GUI. :wink:

------------
Also, if you're interested, [here's what I'm thinking](https://github.com/lah7/polychromatic/commit/dd29d85e5ebbfa7f788ec2742e2de6316d28488e) for detecting devices in the GUIs.

* Detects the ID.
* GUI populates the controls that this component can use.
* When the GUI wants to change something, it is passed to its driver function.

**Comment by terrycain** [comment]
Have read it will reply properly tonight

**Comment by terrycain** [comment]
Ok, yeah I was thinking I would resort to that autostart thing. Will look into it (currently away from all razer devices for about another week so no development until then). Another issue would be the tray applet might start before the Daemon.

That error is evil and is a _known issue_, not a clue how to force it to appear but I think I know a fix for it.

Will look into the "freeze", what do you mean by freeze? Does the entire daemon hang and become unresponsive and require a restart?

If you look at the DBus for the new daemon there is a method that returns device serials. I will eventually abstract this to return you a class representing each device and from there will be functions to do effects etc...

**Comment by lah7** [comment]
> Another issue would be the tray applet might start before the Daemon.

Good point. If that happens, the tray applet could retry a few times within a time interval before it "gives up", I guess?

> ... what do you mean by freeze?

The effect on the keyboard just stopped - like it had frozen in time. In this case, the ripple "paused" but the daemon didn't show any signs it crashed. Setting it to **wave** did not change any lights either. 

> Does the entire daemon hang and become unresponsive and require a restart?

No, it was still running - setting it to **static** allowed the wave/ripple effects to work again. :confused:

**Comment by terrycain** [comment]
Hmm OK, well least the daemon isn't hanging, that's good. Do the logs give anything interesting? Also is it repeatable :wink:, as there's a few places it could go wrong but least it sounds like a logic error.

> the tray applet could retry a few times within a time interval

Yeah, to be fair this is a more than adequate solution. Does the new controller have logging?

**Comment by lah7** [comment]
Oops, I deleted `~/.razer-service`... I think I forgot to put on the verbose flag, so/but I didn't see anything.

> Does the new controller have logging?

Polychromatic Controller? Nope, but I am thinking of adding a `-v` (`--verbose`) and `--vv` switch. I'm not fond of logging the Controller or Tray applet, seems unnecessary. The terminal output will be sufficient for testing and debugging.

I suppose the new daemon library could use its own colour (green :wink:) to differentiate stdout in the terminal? Currently it reads `[DBUS] blah blah`.

**Comment by terrycain** [comment]
No worries.

Yeah, sorry I meant logging to screen, but use the logging module as its so nice (especially the dynamic logging namespaces (like what I do in the daemon)).

And I could change the colour, should be quite simple.

**Comment by terrycain** [comment]
Ooohh would you mind testing the the volume and media keys when you've enabled macrokeys (via dbus).

**Comment by lah7** [comment]
Sure, so a before and after test...

### Before
_Got key error._ But multimedia keys still worked anyway. Same behaviour as Daemonv1.
```
2016-06-04 10:56:19 | razer.device0.keymanager       | ERROR    | Got key error. Couldn't convert event to key name
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/razer_daemon/misc/key_event_management.py", line 367, in key_action
    key_name = EVENT_MAPPING[key_id]
KeyError: 115
```

### After

These packages needed to be installed:

* xautomation
* xdotool

And then the macro keys and multimedia keys worked! Awesome - thanks! :smile:

### Errors caused by missing packages
If it helps, these are the errors that appeared when the packages were not installed:

```
Exception in thread Thread-5:
Traceback (most recent call last):
  File "/usr/lib/python3.5/threading.py", line 914, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3/dist-packages/razer_daemon/misc/key_event_management.py", line 636, in run
    proc = subprocess.Popen(['xdotool', 'key', self._media_key], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
  File "/usr/lib/python3.5/subprocess.py", line 947, in __init__
    restore_signals, start_new_session)
  File "/usr/lib/python3.5/subprocess.py", line 1541, in _execute_child
    raise child_exception_type(errno_num, err_msg)
FileNotFoundError: [Errno 2] No such file or directory: 'xdotool'
```

```
2016-06-04 11:00:44 | razer.device0.keymanager       | INFO     | Running Macro ENTER:[RETURN DOWN, RETURN UP, RIGHTSHIFT DOWN, T DOWN, T UP, RIGHTSHIFT UP, H DOWN, E DOWN, SPACE DOWN, H UP, E UP, SPACE UP, M DOWN, M UP, A DOWN, A UP, C DOWN, R DOWN, C UP, R UP, O DOWN, O UP, SPACE DOWN, SPACE UP, K DOWN, K UP, E DOWN, Y DOWN, E UP, SPACE DOWN, Y UP, SPACE UP, D DOWN, D UP, O DOWN, O UP, E DOWN, S DOWN, E UP, S UP, M1 DOWN, M1 UP, M5 DOWN, M5 UP]
Exception in thread Thread-35:
Traceback (most recent call last):
  File "/usr/lib/python3.5/threading.py", line 914, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3/dist-packages/razer_daemon/misc/macro.py", line 161, in run
    proc = subprocess.Popen(['xte'], stdin=subprocess.PIPE)
  File "/usr/lib/python3.5/subprocess.py", line 947, in __init__
    restore_signals, start_new_session)
  File "/usr/lib/python3.5/subprocess.py", line 1541, in _execute_child
    raise child_exception_type(errno_num, err_msg)
FileNotFoundError: [Errno 2] No such file or directory: 'xte'
```

### Macros
Overall, they work great! There is a bug -- I tried to bind `/gamemode 1` to a macro, but it doesn't like slashes:
```
2016-06-04 11:10:00 | razer.device0.keymanager       | INFO     | Running Macro M5:[FORWARDSLASH DOWN, FORWARDSLASH UP, G DOWN, A DOWN, G UP, A UP, M DOWN, M UP, E DOWN, E UP, M DOWN, M UP, O DOWN, O UP, D DOWN, E DOWN, D UP, E UP, SPACE DOWN, SPACE UP, 1 DOWN, 1 UP]
Unable to resolve keysym for '/'
Unable to resolve keysym for '/'
Unable to resolve keysym for '" "'
Unable to resolve keysym for '" "'
2016-06-04 11:10:00 | razer.device0.macroM5          | DEBUG    | Finished running macro M5
```

I can't type `?` without this happening: `GAMEMODE !`
```
?2016-06-04 11:12:27 | razer.device0.keymanager       | INFO     | Running Macro FORWARDSLASH:[G DOWN, A DOWN, G UP, M DOWN, A UP, E DOWN, M UP, E UP, M DOWN, M UP, O DOWN, O UP, D DOWN, E DOWN, SPACE DOWN, D UP, E UP, SPACE UP, 1 DOWN, 1 UP, M1 DOWN, M1 UP, M1 DOWN, M1 UP]
Unable to resolve keysym for '" "'
Unable to resolve keysym for '" "'
2016-06-04 11:12:27 | razer.device0.macroFORWARDSLASH | DEBUG    | Finished running macro FORWARDSLASH
```

I did try to recording a macro and thought pressing the macro again finished it (my fault, confusing it with the earlier **MacroW**). I like how the daemon does it... is creating macros the same behaviour as Razer's drivers on Windows?
```
2016-06-04 10:59:11 |  razer.device0.keymanager       | WARNING  | Skipping macro assignment as would cause recursion
```

I did notice the macro key itself was lightly "flickering" when it was in a solid state?

Found a nasty bug -- stops me from typing -- trying to deactivate the macro recording after pressing the macro key but not another:

FN+M (activate) .... M1 .... FN+M (deactivate)

```
Exception in thread Thread-2:
Traceback (most recent call last):
  File "/usr/lib/python3.5/threading.py", line 914, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3/dist-packages/razer_daemon/misc/key_event_management.py", line 180, in run
    self._parent.key_action(date, key_code, True)
  File "/usr/lib/python3/dist-packages/razer_daemon/misc/key_event_management.py", line 453, in key_action
    self.add_kb_macro()
  File "/usr/lib/python3/dist-packages/razer_daemon/misc/key_event_management.py", line 496, in add_kb_macro
    start_time = self._current_macro_combo[0][0]
IndexError: list index out of range
```

I know you didn't ask me to test the macro keys, but there you go anyway.

-----------
I have a question... the default effect on the old daemon... (I like to think as a "flow of colours") -- Is this  integrated / possible for the daemon?

**Comment by terrycain** [comment]
I was going to ask you to test the macro keys eventually :wink:. Haven't added `xautomation` and `xdotool` to the Debian control file yet.

> Unable to resolve keysym for '/'

Fixed that a while ago, just haven't updated debs.

I believe its creating macros the same as windows but never used it so not sure (I had to implement all the key logic manually). The Macro LED should flash before you've selected what key the macro should be bound to, then whilst its recording keypresses it should go solid.

> FN+M (activate) .... M1 .... FN+M (deactivate)

Yeah I knew that would be an issue, just haven't been bothered to fix it (easy fix though). Does it output your macro keys to whatever program is in focus whilst your recording a macro? 

> I have a question... the default effect on the old daemon... (I like to think as a "flow of colours") -- Is this integrated / possible for the daemon?

It's not in it yet. Eventually I want an interface which would allow to add this effect and all sorts of similar ones but wont be thought about until I've got the features I currently have to a stable state. But in theory the daemon should be more than capable. Essentially the ripple effect is a custom effect so I just need a way of generalising that.

Glad its working well-ish. It will need some thorough testing, I've started unit-testing some bits too. My plan will be to essentially make my repository the upstream repository and pez's one a "fork" as hes been inactive for too long and there's some repo features I want which I can't access without  being owner so was thinking just leaving a message at the top of his README file to point to the new repo aka mine. Then I can add continuous integration which will test driver compilation, and automatically build deb packages, and I can also add different issue lables etc...

**Comment by lah7** [comment]
> Does it output your macro keys to whatever program is in focus whilst your recording a macro?

Yep, that's what happens and I believe should remain as the behaviour.

Just found another quirk -- I <kbd>FN</kbd> + <kbd>M</kbd> and started typing first (not specifying a macro key) -- every `t` now outputs `tes` (test), even after CTRL+C'd the daemon. :confused: (Re-plugging stopped that)

----------
If you can, I think it might be better to contact the owner to transfer ownership of the original repo to you, it also means the watchers and stars remain on the upstream repo to track progress.

It's all coming together very nicely though. :+1:

**Comment by terrycain** [comment]
Yeah currently you can record a macro on any key (which I believe you can do in synapse. Macro meaning any function not just key combos). If you use the dbus to assign macros (not completed yet) it also opens urls in browsers etc...

What I might do is restrict on the fly macro recording to M1-5. And add in if you FN+M macro_key FN+M it clears macro

**Comment by terrycain** [comment]
Also I've attempted contacting Tim the owner and no response hence the slightly more ugly solution

**Comment by terrycain** [comment]
PC is back up and running :smile: up for another daemon test? :wink: in preperation for its first (possilby heavily buggy) release.

**Comment by lah7** [comment]
Glad to hear! Sure, I can give the daemon another test run. :video_game: 

Are you planning to release the new Python API initially too?

-- :thought_balloon:  I was thinking about this the other day,  maybe tracking what devices are up to could be done at driver level by writing a JSON file somewhere here: `/run/user/1000/somewhere_here.json`?

Did you hear from Tim on the repo transfer?

**Comment by terrycain** [comment]
> Are you planning to release the new Python API initially too?

Yeah, it'll be rough and subject to change (not much change in how it works more on method names etc...) Basically I decided to use DBus's interfaces for what they are actually used for (mostly). so e.g. if the device has a `razer.device.lighting.brighness` interface it will definitely contain both the `getBrightness` and `setBrightness` methods, so am writing a wrapper around that effectively. Will see what you think once its done. Still not quite sure how I want the python interface library but I thought that theres no point constantly thinking about it, ill start coding and if it needs to be changed later on it will need to be changed.

> I was thinking about this the other day, maybe tracking what devices are up

What do you mean by that ha


> Did you hear from Tim on the repo transfer?

And no so I decided to just move to my repo. I'll still push back any driver related changes but that'll be all.

**Comment by terrycain** [comment]
Also have decided not to include anything to "daemonise" the dameon. People will have to run it in a terninal (in super verbose mode :wink:) until the next official release, this way we can iron out any bugs

**Comment by lah7** [comment]
> What do you mean by that

We had [some discussion in May](https://github.com/lah7/razer-chroma-drivers/issues/19#issuecomment-222325564) about tracking the last effect, hence the "what the devices are up to". Here's a refresher:

> The keyboard isn't capable of remembering the current effect.

> I could add in some code to remember the last active effect but it will get out of sync if someone activates effects via the driver directly

I actually forgot about the last bit there about going out of sync. :disappointed:  As all this new stuff is user space, the driver's Python API could possibly write to a file like `/run/user/1000/razer-drivers.json` and store data about it's last actions.

*  *Unix's strengths is reading/writing files. :wink:
* It's `tmpfs`, so it'll always be gone at boot.
* Would need to get user's ID, but `/run/user/<ID>/` is read/writable without being root.

It could be as simple as this:
```
"0000:1111:2222:3333" {
    "type": "keyboard",
    "name": "BlackWidow Chroma"
    "brightness": 255,
    "effect": "custom",
    "color_1": [0,255,0],
    "color_2": [255,0,0]
}
```

If you think this is better handled for the application itself (Polychromatic), then I'll track it there instead. :smile:

**Comment by terrycain** [comment]
> What do you mean by that

Ahhh I remember now, hmmm, I'll have a think on that. Ideally the best place for it would be in the driver. I do have some early code before I knew how to read the brightness where it saved some state, but I'll have to look into that as I would have to have a format in the driver to store things and json may or may not be easy. Worst case I could (with moderate ease) fit it into the daemon at least.

Ideally I don't want to worry about effect tracking until I have the daemon out and working for a bit. :wink: (technically the blackwidow2013 can tell the difference between static and breathing but thats a sideeffect of them bodging the firmware). Later this week ill draw up a proper road map of what i'm aiming to achive before I mark this project at v1. The driver i've already bumped to v1 as thats nice and stable'ish.

**Comment by terrycain** [comment]
@lah7 if your interested, join that telegram chat thingy, am just about to start planning python lib api with the other guy who made an app. You input would be welcome and most appreciated.
